### ISSUE TITLE
Some doc is out dated

### ISSUE THREAD

**Issue by ailisp** [mmontone/cl-rest-server#4]
I noticed the `implement-resource-operation` is now different as doc said. For example
```lisp
(implement-resource-operation get-user (id &key expand)
   (serialize (find-user id) :expand expand))
```
should change to
```lisp
(implement-resource-operation apt-test get-user (id &key expand)
   (serialize (find-user id) :expand expand))
```

**Comment by mmontone** [comment]
Yes. Sorry. I'll try to update the docs when I find time. The library has gone through some changes.

I was not aware there were users of the library, apart from me. :)

**Comment by mmontone** [comment]
I guess your best bet is to look at test code for now until I get to update the docs.

**Comment by ailisp** [comment]
Thanks! You are so kind to say there are no users, I think rest-server
provide the most declarative and powerful REST api define tools. I have
read the code and tests, it now works well.

2017年2月27日 下午8:03，"Mariano Montone" <notifications@github.com>写道：

**Comment by mmontone** [comment]
I wish it worked better!

Some of the things I tried didn't work out, like having transparent JSON
and XML serialization and de-serialization. It works for JSON only for now.

Also the dispatching system is a bit messy. Could be less messy and also
more complete, webmachine like.

I'm not kind when I say there are no users, I'm really not aware of any
users :)

Cheers

**Comment by ailisp** [comment]
For dispatching, are you mean [this]( https://github.com/webmachine/webmachine/wiki/Dispatching)
And if when define api:
```lisp
(define-api api-test
    (:title "Api test"
            :documentation "This is an api test")
  (parameters (:produces (:json)
                         :consumes (:json)
                         :documentation "Parameters test"
                         :path "/parameters")
              (parameters (:produces (:json)
                                     :consumes (:json)
                                     :documentation "Parameters test"
                                     :path "/parameters")
                          (&optional (boolean :boolean nil "A boolean parameter")
                                     (integer :integer nil "An integer parameter")
                                     (string :string nil "A string parameter")
                                     (list :list nil "A list parameter")))) ...)
```

If the inside `parameters` can derive options from the outside `parameters` that not provided it will be more convenient. Or we may use:
```lisp
(define-api api-test
    (:title "Api test"
            :documentation "This is an api test")
  (parameters (:produces (:json)
                         :consumes (:json)
                         :documentation "Parameters test"
                         :path "/parameters")
              (:default
                          (&optional (boolean :boolean nil "A boolean parameter")
                                     (integer :integer nil "An integer parameter")
                                     (string :string nil "A string parameter")
                                     (list :list nil "A list parameter")))) ...)
```

to indicate the same. This only my personal preference, and not as important as XML serialization and dispatching. If you think these change (in define-api) is reasonable, I can send you a pull request ;)

**Comment by mmontone** [comment]
No, I mean this: https://github.com/mmontone/cl-rest-server/blob/master/src/api.lisp#L171

I'm not happy with that code.

Also, it doesn't implement this: https://raw.githubusercontent.com/webmachine/webmachine/develop/docs/http-headers-status-v3.png

Like other frameworks out there.

https://github.com/inhabitedtype/ocaml-webmachine

**Comment by ailisp** [comment]
It's quite a lot error handling to do, wish cl-rest-server can be mature enough in the future

**Comment by mmontone** [comment]
The implementation technique for that is "parsing" that diagram in some way (you could have a declarative spec of that diagram) and generate the conditional code that does the dispatching. Sounds like fun.

**Comment by ailisp** [comment]
Sounds like something I learned from Software Engineering course of auto generate Java program from UML. The program generated is 90% accurate, but that's enough. It's cool if you implement one for Common Lisp.
